core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,201 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+import com.mesosphere.usi.models._
+
+/**
+  * Container class responsible for keeping track of the frame and cache.
+  *
+  * As a general rule, business logic does not manipulate the USI state directly, but rather does so by returning
+  * effects.
+  */
+private[core] class SchedulerLogic {
+  private var frame: Frame = Frame(Map.empty, Map.empty, Map.empty)
+  private var cachedPendingLaunch = CachedPendingLaunch(Set.empty)
+
+  def processSpecEvent(msg: SpecEvent): FrameEffects = {
+    val (frameWithPodSpecChanges, changedPodIds) = frame.applySpecEvent(msg)
+    frame = frameWithPodSpecChanges
+    val frameEffects = SchedulerLogic.computeNextStateForPods(frameWithPodSpecChanges)(changedPodIds)
+    applyAndUpdateCaches(frameEffects, changedPodIds)
+  }
+
+  /**
+    * Given a frame effects, apply the usi state change effects of the frame effects. Then, update the cache and perform
+    * other housekeeping items for pods that are dirty, such as:
+    *
+    * - Prune terminal / unreachable podStatuses for which no podSpec is defined
+    * - Update the pending launch set index
+    * - (WIP) issue any revive calls (this should be done elsewhere)
+    *
+    * Any usiState changes (such as pruning terminal / unreachable podStatuses) are added to the frameEffects.
+    *
+    * @param eventFrameEffects The effects from the incoming SpecEvent
+    * @param dirtyPodIds The podSpecs whose cache should be updated or checked for pruning
+    * @return The frame effects with additional housekeeping effects applied
+    */
+  private def applyAndUpdateCaches(eventFrameEffects: FrameEffects, dirtyPodIds: Set[PodId]): FrameEffects = {
+    frame = frame.applyStateEffects(eventFrameEffects)
+    this.cachedPendingLaunch = this.cachedPendingLaunch.update(frame, dirtyPodIds)
+
+    val pruneEffects = SchedulerLogic.pruneTaskStatuses(frame)(dirtyPodIds)
+    frame = frame.applyStateEffects(pruneEffects)
+
+    val reviveEffects =
+      if (cachedPendingLaunch.pendingLaunch.nonEmpty)
+        FrameEffects.empty.withMesosCall(Mesos.Call.Revive)
+      else
+        FrameEffects.empty
+
+    eventFrameEffects ++ pruneEffects ++ reviveEffects
+  }
+
+  /**
+    * Process a Mesos event and update internal state.
+    *
+    * @param event
+    * @return The events describing state changes as Mesos call intents
+    */
+  def processMesosEvent(event: Mesos.Event): FrameEffects = {
+    val result = SchedulerLogic.handleMesosEvent(frame, cachedPendingLaunch.pendingLaunch)(event)
```

Comment:
This split into `case class SchedulerLogic` and `object SchedulerLogic` is only necessary (to ease the testing) because `SchedulerLogic` holds the `frame`. If the frame would live outside the both could be united into a set of pure methods: _get a frame and event -> return new frame and effects_. 
I believe @jeschkies suggested something similar and I support that.

Reply:
Added FrameWithEffects abstraction which handles this more elegantly. Preference to have business logic only return effects if possible.